### ISSUE TITLE
C++17 standard deprecates std::iterator

### ISSUE THREAD

**Issue by jefferis** [google/draco#966]
I have wrapped draco in a simple read-only library for the R language, [dracor](https://github.com/natverse/dracor) available from the main [CRAN package repository](https://cran.r-project.org/web/packages/dracor/index.html)).

Unfortunately I have just received a warning that dracor will be removed from CRAN by 2023-02-26 unless I can fix a new compiler warning associated with a switch to C++17 as their standard compiler. Specifically:

```
Found the following significant warnings:
  draco/mesh/corner_table_iterators.h:25:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  draco/mesh/corner_table_iterators.h:115:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  draco/mesh/corner_table_iterators.h:197:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  ./draco/mesh/corner_table_iterators.h:25:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  ./draco/mesh/corner_table_iterators.h:115:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  ./draco/mesh/corner_table_iterators.h:197:19: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
used C++ compiler: ‘g++-12 (Debian 12.2.0-14) 12.2.0’
```
See https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-debian-gcc/dracor-00install.html for further details.

This seems to be a fairly well-known issue ([discussed in some detail here](https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/)) but the solution is not completely trivial. I wondered if anyone is actually working or likely to work on addressing this in the draco codebase any time soon @tomfinegan? It seem that all that is required is [some extra boiler plate](https://stackoverflow.com/questions/37031805/preparation-for-stditerator-being-deprecated) that would be both backwards and forwards compatible, but my understanding of the details are poor. Thanks so much!

**Comment by jefferis** [comment]
I think for [VertexRingIterator](https://github.com/google/draco/blob/5c0976b219b3f05b8935893ff38249b4db6088f5/src/draco/mesh/corner_table_iterators.h#L22-L30)
```diff
 template <class CornerTableT>
-class VertexRingIterator
-    : public std::iterator<std::forward_iterator_tag, VertexIndex> {
+class VertexRingIterator {
  public:
+   // iterator traits
+   using value_type = VertexIndex;
+   using difference_type = std::ptrdiff_t;
+   using pointer = const VertexIndex*;
+   using reference = const VertexIndex&;
+   using iterator_category = std::forward_iterator_tag;
+
   // std::iterator interface requires a default constructor.
   VertexRingIterator()
       : corner_table_(nullptr),
```
the change would look something like this but I'm not too sure about the iterator traits e.g. `difference_type`.

**Comment by ondys** [comment]
Thanks for the report. We actually have this fixed internally already and it will be included in the next release.

**Comment by jefferis** [comment]
Thank you @ondys, that's great. Do you have a (rough) timeline for that release?

**Comment by ondys** [comment]
The plan is to do it either this week or early next week.

**Comment by jefferis** [comment]
Thanks @ondys. For my reference this was fixed in [v1.5.6](https://github.com/google/draco/pull/972/commits/9880f9b67897f617f87d94fca50542800ee7a044#diff-92c93ec0f13782e277d00c43d1c4ae9fdc90f1f3430137692246062ee00dcb31)
